### DISCUSSION TITLE
aws cognito integration

### DISCUSSION THREAD

**Discussion by scorcism** [supabase/supabase#28923] [Questions]
I have followed all teh steps here `https://supabase.com/docs/guides/auth/third-party/aws-cognito?queryGroups=cognito-create-client&cognito-create-client=ts`. 
I guess there should be more details and informations about what others things to do to integrate thsi will the supabase db.
> If this is already considered, aplogies, plss suggest docs.

Where am I ?
I am able to get the response from the `https://${process.env.REACT_APP_COGNITO_DOMAIN}/oauth2/token` but after this what should I have to continue on to get the supabase sessions working with the current logged in user.

**Comment by rynvelt** [comment]
I am also encountering problems when using Realtime with Cognito as an external identity provider which prevent me from using Supabase at the moment.
I might list what I have found out so far as this might help others.

I have managed to setup Supabase with my own Cognito pool using the [react-oidc-context](https://github.com/authts/react-oidc-context) library in my React application. It looks something like this:

```ts
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./index.css";
import App from "./App.tsx";

import { AuthProvider, type AuthProviderProps } from "react-oidc-context";
import { ReactQueryClientProvider } from "./components/ReactQueryClientProvider.tsx";

const cognitoAuthConfig: AuthProviderProps = {
	authority: import.meta.env.VITE_COGNITO_USER_POOL_DOMAIN,
	client_id: import.meta.env.VITE_COGNITO_CLIENT_ID,
	redirect_uri: import.meta.env.VITE_COGNITO_REDIRECT_URI,
	post_logout_redirect_uri: import.meta.env.VITE_COGNITO_LOGOUT_URI,
	response_type: "code",
	scope: "phone openid email",
	automaticSilentRenew: true,
};

const root = createRoot(document.getElementById("root")!);
root.render(
	<StrictMode>
		<ReactQueryClientProvider>
			<AuthProvider {...cognitoAuthConfig}>
				<App />
			</AuthProvider>
		</ReactQueryClientProvider>
	</StrictMode>,
);
```

And then using a hook:

```ts
import { createClient } from "@supabase/supabase-js";
import type { SupabaseClient } from "@supabase/supabase-js";
import { useEffect, useState } from "react";
import { useAuth } from "react-oidc-context";
import type { Database } from "../generated/database";

export type TypedSupabaseClient = SupabaseClient<Database>;

export function useSupabaseClient(): TypedSupabaseClient | null {
	const auth = useAuth();
	const [client, setClient] = useState<TypedSupabaseClient | null>(null);
	useEffect(() => {
		const accessToken = auth.user?.access_token ?? null;
		const load = async () => {
			const supabaseClient = createClient(
				import.meta.env.VITE_SUPABASE_URL,
				import.meta.env.VITE_SUPABASE_ANON_PUBLIC_KEY,
				{
					realtime: {
						accessToken: async () => {
							return accessToken;
						},
						timeout: 3000,
						reconnectAfterMs: () => 3000,
					},
					accessToken: async () => {
						return accessToken;
					},
				},
			);
			setClient(supabaseClient);
		};
		load();
	}, [auth]);
	return client;
}
```

And then using the hook like this:

```ts
	const supabase = useSupabaseClient();
	const getCountries = useCallback(async () => {
		if (!supabase) {
			return [];
		}
		const { data } = await supabase.from("countries").select();
		return data;
	}, [supabase]);

	const { data: countries } = useQuery({
		queryKey: ["countries"],
		queryFn: getCountries,
	});
```

You need to fiddle around with the token that Cognito generates because Supabase seems to expect a "role" claim with value "authenticated", so you have to set up a Lambda function like this and then use it as a [pre-token-generation](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html) function (you have to create the Lambda and then link it in your Cognito user pool such that is triggered upon token generation):

```js
export const handler = (event, context) => {
  event.response = {
      claimsAndScopeOverrideDetails: {
          idTokenGeneration: {
              claimsToAddOrOverride: {
                role: "authenticated"
              },
          },
          accessTokenGeneration: {
              claimsToAddOrOverride: {
                  role: "authenticated"
              },
          },
      }
  };
  context.done(null, event);
};
```

This is well described in the Supabase [documentation](https://supabase.com/docs/guides/database/extensions/wrappers/cognito).
The Access Token I get from Cognito works when used with the Supabase HTTP API but it does not work with Realtime for me. With Realtime I see an error on the Websocket channel in my Browser shortly after the Websocket connection has been established and the error message looks like this:

```json
{
  "ref": null,
  "event": "system",
  "payload": {
    "message": ":error_generating_signer",
    "status": "error",
    "extension": "system",
    "channel": "REALTIME_COUNTRIES"
  },
  "topic": "realtime:REALTIME_COUNTRIES"
}
```

I can only assume that this has something to do with the token from Cognito, because if I use the almighty service_role JWT (which is allowed to do anything) from the Supabase Dashboard instead of the JWT from Cognito my code suddenly works with Realtime.

I would appreciate if it would be documented better which assumptions are made by Supabase about the JWTs from external providers and if there are any differences between the regular HTTP API and Realtime. I compared the JWT from Cognito with the one from Supabase and I noticed the following:

- The Access Token JWT from Cognito is signed by asymmetric methods (RS256) while the Supabase service_role JWT is signed using a symmetric method (HS256). There seem to be some discussions about introducing asymmetric signing of JWTs (see https://github.com/orgs/supabase/discussions/12759). I'm not sure if this only affects creating JWTs when using Supabase Auth (instead of an external provider) or if this issue also is related to current authentication mechanisms in Realtime
- The Access Token JWT from Cognito seems to lack an aud claim (audience). This is intended as according to the [AWS documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html) and it is possible to set an aud claim, but only with a very specific value equal to the clientId. The [documentation](https://supabase.com/docs/guides/auth/auth-hooks/custom-access-token-hook) on custom AccessToken hooks however lists the aud claim as a mandatory field, among some other fields

If somebody from Supabase could maybe clarify or give a hint to what the issue might be here I would really appreciate that.

I know that there are also some issues on Github related to how to setup the access token properly for Realtime. There is also the option to set it like this, which is described here and surely somewhere else [here](https://supabase.com/docs/guides/realtime/postgres-changes#custom-tokens):

```typescript
 const { user, error } = supabase.auth.setAuth(access_token)
```

I have also tried setting it with this method, with the same result, so I just assume that Supabase does not l ike the JWT access token from Cognito.

**Reply by rynvelt** [reply]
Just reading this [issue](https://github.com/supabase/supabase-js/issues/553#issuecomment-1573518898) and trying to understand what people did. I might try give it a try.

**Reply by rynvelt** [reply]
It might be the case that the Realtime authentication checks for the field "typ": "JWT" in the JWT header as described by another user [here](https://github.com/supabase/realtime/issues/562#issuecomment-1490207303) and the [code](https://github.com/supabase/realtime/blob/61f4930a0576ce2f3b67643a5a260418fbb75352/lib/realtime_web/channels/auth/jwt_verification.ex#L55).

However, Cognito access tokens do not have a "typ" header field, see [here](https://repost.aws/questions/QUdDANuP-mSLaNNwpQevEsyQ/cognito-oauth-access-token-missing-typ-header-parameter) and [AWS Cognito Access Token documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-access-token.html).

For me this means that it is likely that Realtime will not work with Cognito - it would be helpful if the Supabase docs could tell me that on their [Cognito integration documentation site](https://supabase.com/docs/guides/auth/third-party/aws-cognito?queryGroups=cognito-create-client&cognito-create-client=ts). 

If anyone from the Supabase team could just clarify or confirm whether Realtime works with Cognito I would be thankful. Or maybe someone here has had more success than me and wants to share?
